File: Bronze/2018/December/03_back_and_forth/BackAndForth.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.PrintWriter;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.StringTokenizer;

/**
 * @author Red
 * @version 1.0
 */
public class BackAndForth {
    static List<Integer> buckets1 = new ArrayList<Integer>();
    static List<Integer> buckets2 = new ArrayList<Integer>();
    static HashSet<Integer> milks = new HashSet<Integer>();

    public static void main(String[] args) throws Exception {
        BufferedReader r = new BufferedReader(new FileReader("backforth.in"));
        PrintWriter pw = new PrintWriter("backforth.out");
        StringTokenizer st = new StringTokenizer(r.readLine());
        for (int i = 0; i < 10; i++) {
            buckets1.add(Integer.parseInt(st.nextToken()));
        }
        st = new StringTokenizer(r.readLine());
        for (int i = 0; i < 10; i++) {
            buckets2.add(Integer.parseInt(st.nextToken()));
        }
        move(new ArrayList<>(), buckets1, buckets2);
        pw.println(milks.size());
        pw.close();
    }

    /**
     * 递归移动
     * Java 语言中引用数据类型（非基本数据类型）均为引用传递
     * 在函数内用 copy 的数据传递给下一层可以达到本层的回溯效果
     */
    public static void move(List<Integer> milkMount, List<Integer> bucketsFrom, List<Integer> bucketsTo) {
        if (milkMount.size() == 4) {
            int mount = 1000;
            mount -= (milkMount.get(0) + milkMount.get(2));
            mount += (milkMount.get(1) + milkMount.get(3));
            milks.add(mount);
            return;
        }
        for (int i = 0; i < bucketsFrom.size(); i++) {
            List<Integer> bucketsFromCopy = new ArrayList<>(bucketsFrom);
            List<Integer> bucketsToCopy = new ArrayList<>(bucketsTo);
            List<Integer> milkMountCopy = new ArrayList<>(milkMount);
            int d = bucketsFromCopy.remove(i);
            bucketsToCopy.add(d);
            milkMountCopy.add(d);
            move(milkMountCopy, bucketsToCopy, bucketsFromCopy);
        }
    }
}

File: Bronze/2018/December/03_back_and_forth/back_and_forth.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <algorithm>
#include <unordered_set>

using namespace std;

vector<int> buckets1;
vector<int> buckets2;
unordered_set<int> milks;
// 递归移动
void move(vector<int> milk_mount, vector<int>& buckets_from, vector<int>& buckets_to) {
    if (milk_mount.size() == 4) {
        int mount = 1000;
        mount -= (milk_mount[0] + milk_mount[2]);
        mount += (milk_mount[1] + milk_mount[3]);
        milks.insert(mount);
        return;
    }
    for (int i = 0; i < buckets_from.size(); i++) {
        vector<int> buckets_from_copy = buckets_from;
        vector<int> buckets_to_copy = buckets_to;
        vector<int> milk_mount_copy = milk_mount;
        int d = buckets_from_copy[i];
        buckets_from_copy.erase(buckets_from_copy.begin() + i);
        buckets_to_copy.push_back(d);
        milk_mount_copy.push_back(d);
        move(milk_mount_copy, buckets_to_copy, buckets_from_copy);
    }
}

int main() {
    ifstream fin("backforth.in");
    ofstream fout("backforth.out");

    for (int i = 0; i < 10; i++) {
        int bucket;
        fin >> bucket;
        buckets1.push_back(bucket);
    }

    for (int i = 0; i < 10; i++) {
        int bucket;
        fin >> bucket;
        buckets2.push_back(bucket);
    }

    move({}, buckets1, buckets2);

    fout << milks.size() << endl;

    fin.close();
    fout.close();

    return 0;
}
